src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleReference.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System.Collections.Generic;
+using System.Reflection;
+using System.Text.Json.Serialization;
+
+namespace System.Text.Json
+{
+    public static partial class JsonSerializer
+    {
+        private enum ResolvedReferenceHandling
+        {
+            Ignore,
+            Preserve,
+            //Ignore or Error were selected but no loop was found.
+            None
+        }
+
+        private static ResolvedReferenceHandling ResolveReferenceHandling(JsonSerializerOptions options, ref WriteStack state, out int referenceId, out bool writeAsReference, object currentPropertyValue = null)
+        {
+            ReferenceHandlingOnSerialize handling;
+
+            //This might be a bit expensive...
+            //Property has attribute?
+            JsonReferenceHandlingAttribute attr = JsonPropertyInfo.GetAttribute<JsonReferenceHandlingAttribute>(state.Current.JsonPropertyInfo?.PropertyInfo);
+            if (attr != null)
+            {
+                handling = attr.Handling;
+            }
+            //Class has attribute?
+            else if ((attr = state.Current.JsonClassInfo.Type.GetCustomAttribute<JsonReferenceHandlingAttribute>()) != null)
+            {
+                handling = attr.Handling;
+            }
```

Comment:
We might consider dropping attribute support... this is causing a huge performance decrease on the main path (around 80% increase in runtime).

@ahsonkhan @steveharter thoughts?